ament_cpplint/ament_cpplint/main.py

```diff
@@ -144,27 +149,49 @@ def main(argv=sys.argv[1:]):
         else:
             print("Not using '--root'")
         print('')
-        arguments += files
-        filenames = ParseArguments(arguments)
-
-        for filename in filenames:
-            # hook into error reporting
-            errors = []
-
-            def custom_error(filename, linenum, category, confidence, message):
-                if ament_cpplint.cpplint._ShouldPrintError(category, confidence, linenum):
-                    errors.append({
-                        'linenum': linenum,
-                        'category': category,
-                        'confidence': confidence,
-                        'message': message,
-                    })
-                DefaultError(filename, linenum, category, confidence, message)
-            ament_cpplint.cpplint.Error = custom_error
-
-            ProcessFile(filename, _cpplint_state.verbose_level)
-            report.append((filename, errors))
-            print('')
+
+        files_to_avoid = []
+        for f in files:
+            path, file_name = os.path.split(f)
+            path_to_check = pathlib.Path(path)
+
+            def check_path_parents(parents, exclude):
+                for p in parents:
+                    if p.match(exclude):
+                        return True
+                return False
+
+            for excl in args.excludes:
+                if (
+                    excl == file_name or
+                    path_to_check.match(excl) or
+                    check_path_parents(path_to_check.parents, excl)
+                ):
+                    files_to_avoid.append(f)
+        files_check = list(set(files) ^ set(files_to_avoid))
+
+        if(len(files_check) != 0):
```

Comment:
This check us only necessary since the previous check from line 123 isn't effective since the exclude logic is applied afterwards. How about moving the logic into `get_file_groups()` - which I think would be the most efficient approach since it even avoid crawling / collecting files under an excluded directory. I would expect that would also avoid the need to check all `parents`.

Nitpick: no parenthesis around Python conditions. The Pythonic way to write this condition would be: `if files_check`.

Reply:
Logic moved to `get_file_groups()` https://github.com/ament/ament_lint/pull/238/commits/3b9b453dc0e4361ff7c94a6f054fd91da8f6d444

Reply:
The `exclude` logic can already be applied to the `dirnames` variable to filter them out early (without the need to recurse into them). That will also simplify the check later which doesn't have to consider parents.

Reply:
moved to dirnames https://github.com/ament/ament_lint/pull/238/commits/3b9b453dc0e4361ff7c94a6f054fd91da8f6d444

Reply:
I don't see where the code filters `dirnames` based on the `excludes`?